src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -1140,9 +1142,280 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, CSharpCompi
 
             return new BoundCall(node, receiver, method, args, argNames, argRefKinds, isDelegateCall: isDelegateCall,
                         expanded: expanded, invokedAsExtensionMethod: invokedAsExtensionMethod,
-                        argsToParamsOpt: argsToParams, resultKind: LookupResultKind.Viable, binderOpt: this, type: returnType, hasErrors: gotError);
+                        argsToParamsOpt: argsToParams, defaultArgumentsOpt, resultKind: LookupResultKind.Viable, binderOpt: this, type: returnType, hasErrors: gotError);
+        }
+
+#nullable enable
+
+        private static SourceLocation GetCallerLocation(SyntaxNode syntax)
+        {
+            switch (syntax.Kind())
+            {
+                case SyntaxKind.InvocationExpression:
+                    return new SourceLocation(((InvocationExpressionSyntax)syntax).ArgumentList.OpenParenToken);
+                case SyntaxKind.ObjectCreationExpression:
+                    return new SourceLocation(((ObjectCreationExpressionSyntax)syntax).NewKeyword);
+                case SyntaxKind.BaseConstructorInitializer:
+                case SyntaxKind.ThisConstructorInitializer:
+                    return new SourceLocation(((ConstructorInitializerSyntax)syntax).ArgumentList.OpenParenToken);
+                case SyntaxKind.PrimaryConstructorBaseType:
+                    return new SourceLocation(((PrimaryConstructorBaseTypeSyntax)syntax).ArgumentList.OpenParenToken);
+                case SyntaxKind.ElementAccessExpression:
+                    return new SourceLocation(((ElementAccessExpressionSyntax)syntax).ArgumentList.OpenBracketToken);
+                default:
+                    return new SourceLocation(syntax.GetFirstToken());
+            }
+        }
+
+        internal BoundExpression BindDefaultArgument(SyntaxNode syntax, ParameterSymbol parameter, Symbol containingMember, bool enableCallerInfo, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(parameter.IsOptional);
+
+            TypeSymbol parameterType = parameter.Type;
+
+            var defaultConstantValue = parameter.ExplicitDefaultConstantValue switch
+            {
+                // Bad default values are implicitly replaced with default(T) at call sites.
+                { IsBad: true } => ConstantValue.Null,
+                var constantValue => constantValue
+            };
+
+            var callerSourceLocation = enableCallerInfo ? GetCallerLocation(syntax) : null;
+            BoundExpression defaultValue;
+            bool allowFailingConversion;
+            if (defaultConstantValue == ConstantValue.Unset)
+            {
+                // This is only expected to occur in recursive error scenarios, for example: `void F(object param = F()) { }`
+                defaultValue = new BoundDefaultExpression(syntax, parameterType) { WasCompilerGenerated = true };
```

Comment:
`parameter.ExplicitDefaultConstantValue` doesn't return a bad constant value in this scenario, because it doesn't really know that something is wrong yet. Using the `void F(object param = F()) { }` case as an example:
1. We compute the default value for `param`. The first thing we do is bind the default syntax value.
2. We bind the call to `F()`, including its default arguments.
3. We compute the default value for `param`. (This is reentrant.)
    - Since the completion part `StartDefaultSyntaxValue` has been set, we just return `ConstantValue.Unset`
4. After that we pass the bound default value (the call `F()`) to `ParameterHelpers.ReportDefaultParameterErrors`.
    - This reports `ERR_DefaultValueMustBeConstant` and returns `true`
5. Since `ReportDefaultParameterErrors` returned `true` we return `ConstantValue.Bad` as the "final" default value for the parameter.

I'll add a test that will hopefully illustrate this better, at least the last part, since I can't seem to find such a test added in this PR already.

Note that the behavior of returning `ConstantValue.Unset` occurs both for reentrant scenarios as well as for multithreaded scenarios. It's possible that we want to do something like have a reentrant usage return `ConstantValue.Bad`, while a multithreaded scenario will spin wait until the constant value is available. I am actually not sure how to accomplish that in `SourceComplexParameterSymbol.DefaultSyntaxValue`. Perhaps the binder should instead avoid accessing the default value on a parameter if `BinderFlags.ParameterDefaultValue` is set and just use `ConstantValue.Bad` as a placeholder.

Also, I had hoped to pass `hasErrors: true` here so that we will always fail in emit if no other (better) diagnostic is given, but it ended causing some diagnostic reporting in `ParameterHelpers.ReportDefaultParameterErrors` to be skipped, and I wasn't sure how to square it all without causing a lot of test churn.